p4_symbolic/symbolic/symbolic.h

```diff
@@ -37,91 +38,37 @@ namespace symbolic {
-// metadata fields, like we do with concrete header. I am using the most
-// flexible option for now, and will be able to better understand this after
-// implementing and experimenting with this a bit.
-using ConcreteMetadata = std::unordered_map<std::string, int>;
-
-// Provides symbolic handles for the fields of the symbolic packet used by
-// our solver. These handles can be used to contstrain the conrete output
-// packets.
-struct SymbolicHeader {
-  // TODO(babman): Swith to bit strings.
-  z3::expr eth_src;   // 48 bit.
-  z3::expr eth_dst;   // 48 bit.
-  z3::expr eth_type;  // 16 bit.
-
-  z3::expr outer_ipv4_src;        // 32 bit, valid if eth_type = 0x0800
-  z3::expr outer_ipv4_dst;        // 32 bit, valid if eth_type = 0x0800
-  z3::expr outer_ipv6_dst_upper;  // 64 bit, valid if eth_type = 0x86dd
-  z3::expr outer_ipv6_dst_lower;  // 64 bit, valid if eth_type = 0x86dd
-  z3::expr outer_protocol;        // 8 bit, valid if eth_type is ip
-  z3::expr outer_dscp;            // 6 bit, valid if eth_type is ip
-  z3::expr outer_ttl;             // 8 bit, valid if eth_type is ip
-
-  z3::expr inner_ipv4_dst;        // 32 bit, valid if outer_protocol = 4
-  z3::expr inner_ipv6_dst_upper;  // 64 bit, valid if outer_protocol = 4
-  z3::expr inner_ipv6_dst_lower;  // 64 bit, valid if outer_protocol = 41
-  z3::expr inner_protocol;        // 8 bit, valid if outer_protocol = 4/41
-  z3::expr inner_dscp;            // 6 bit, valid if outer_protocol = 4/41
-  z3::expr inner_ttl;             // 8 bit, valid if outer_protocol = 4/41
-
-  z3::expr icmp_type;  // 8 bit, valid if eth_type is ip
-  z3::expr vid;        // 12 bit, valid if eth_type = 0x6007
-};
-
-// The symbolic counterpart of ConcreteMetadata. This can be used to constrain
-// intermediate values.
-using SymbolicMetadata = std::unordered_map<std::string, z3::expr>;
+// Maps the name of a header field in the p4 program to its concrete value.
+using ConcreteHeaders = std::unordered_map<std::string, std::string>;
+
+// The symbolic counterpart of ConcreteHeader.
+// Maps the name of a header field in the p4 program to its symbolic value.
+// This can be used to constrain p4 program fields.
+// This is automatically constructred from the header type definitions
+// the p4 program has.
+// Assume the p4 program has a header instance named "standard_metadata" of type
+// "standard_metadata_t", which has field "ingress_port" of type "bit<9>" in it.
+// Then, we will have:
+//     SymbolicMetadata["standard_metadata.ingress_port"] =
+//         <symbolic bit vector of size 9>
+using SymbolicHeaders = GuardedMap;
 
 // Expresses a concrete match for a corresponding concrete packet with a
 // table in the program.
 struct ConcreteTableMatch {
   bool matched;  // false if no entry in this table was matched, true otherwise.
   // if matched is false, these two fields are set to -1.
   int entry_index;
-  int value;
+  std::string value;
```

Comment:
Does value even make sense anymore in your world of actions?

Reply:
If not, we can probably just delete it.

Reply:
Honestly, I am unsure. I have left a couple of TODO comments about this, and I am planning on taking a closer look at this next. I kept it for compatibility with the old Hercules API that we discussed some time ago. I think it can probably go away. But I wanted to understand this a bit better before making the change.

Reply:
I think you should get rid of it. It doesn't make sense anymore.

Reply:
done.